{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pipeline (HGSOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/haz19024/.venv/work/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/haz19024/.local/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3460: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('../ov2295_snv_counts.csv')\n",
    "df = df.astype({'chrom': 'str', \n",
    "                'coord': 'int32', \n",
    "                'ref': 'str',\n",
    "                'alt': 'str',\n",
    "                'ref_counts': 'int32',\n",
    "                'alt_counts': 'int32'})\n",
    "df['snp_id'] = df['chrom'] + ':' + df['coord'].astype(str) + ':' + df['ref'] + ':' + df['alt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "# filter out sites with low coverage\n",
    "# def filtering(df, min_total_counts=0, ):\n",
    "#     df = df[df['total_counts']>min_total_counts]\n",
    "#     return df\n",
    "\n",
    "# false positive: sequencing error, usually bewteen 0.01 to 0.1\n",
    "# false negative: combined error from amplication, sequencing and dropout, we assume 0.2 or even higher\n",
    "# genotype prior: flat prior, gt can be either 0 or 1\n",
    "def posterior_prob_with_fn_fp_flat_ht(ref_counts, alt_counts, fp=0.01, fn=0.2, prior_ref=0.5, margin=1e-5):\n",
    "    # binary case : haplotype\n",
    "    p00, p01, p10, p11 = 1-fp, fp, fn, 1-fn\n",
    "    g0 = prior_ref * (p00**ref_counts) * (p01**alt_counts)\n",
    "    g1 = (1-prior_ref) * (p10**ref_counts) * (p11**alt_counts)\n",
    "    prob =  max(min(np.round(g0/(g0+g1), 5), 1-margin), margin)\n",
    "    return prob\n",
    "\n",
    "def posterior_prob_with_fn_fp_flat_gt(ref_counts, alt_counts, fp=0.01, fn=0.2, prior_ref=0.5, margin=1e-5):\n",
    "    # trinary case : genotype\n",
    "    p00, p01, p10, p11 = 1-fp, fp, fn, 1-fn\n",
    "    l00 = p00**ref_counts * p01**alt_counts\n",
    "    l11 = p10**ref_counts * p11**alt_counts\n",
    "    l01 = (0.5*(p00+p10))**ref_counts * (0.5*(p01+p11))**alt_counts\n",
    "    g00 = prior_ref**2 * l00\n",
    "    g01 = 2*(1-prior_ref)*prior_ref * l01\n",
    "    g11 = (1-prior_ref)**2 * l11\n",
    "    prob =  max(min(np.round(g00/(g00+g01+g11), 5), 1-margin), margin)\n",
    "    return prob\n",
    "\n",
    "\n",
    "def phred_likelihood_with_fn_fp_flat(ref_counts, alt_counts, fp=0.01, fn=0.2):\n",
    "    # Q-phred score = -10 * log10(p)\n",
    "    p00, p01, p10, p11 = 1-fp, fp, fn, 1-fn\n",
    "    l00 = p00**ref_counts * p01**alt_counts\n",
    "    l11 = p10**ref_counts * p11**alt_counts\n",
    "    l01 = (0.5*(p00+p10))**ref_counts * (0.5*(p01+p11))**alt_counts\n",
    "    q00 = int(-10 * np.log10(l00))\n",
    "    q11 = int(-10 * np.log10(l11))\n",
    "    q01 = int(-10 * np.log10(l01))\n",
    "    return q00, q01, q11\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SA921 616\n",
      "SA922 588\n",
      "SA1090 739\n"
     ]
    }
   ],
   "source": [
    "# check the number of cell ids for each sample\n",
    "for sample in df['sample_id'].unique():\n",
    "    print(sample, df[df['sample_id'] == sample]['cell_id'].unique().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter out samples with preclustered cells\n",
    "df_clones = pd.read_csv('../ov2295_clone_clusters.csv')\n",
    "filtered_cells = df_clones['cell_id'].unique()\n",
    "# filtered df\n",
    "df = df[df['cell_id'].isin(filtered_cells)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "891 14068\n"
     ]
    }
   ],
   "source": [
    "# prelimary view on data\n",
    "cells = df['cell_id'].drop_duplicates().tolist()\n",
    "sites = df['snp_id'].drop_duplicates().tolist()\n",
    "cells = sorted(cells)\n",
    "sites = sorted(sites)\n",
    "print(len(cells), len(sites))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # filter out non-informative sites, criterion: more than 650 cells missed\n",
    "# informative_df = df.groupby('snp_id').size().sort_values(ascending=False)\n",
    "# informative_sites = informative_df[informative_df > len(cells) - 650].index.tolist()\n",
    "# df = df[df['snp_id'].isin(informative_sites)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "891 14068\n",
      "ov2295_14068_891\n"
     ]
    }
   ],
   "source": [
    "# check again the number of cells and the number of sites, in HUNTRESS, there are 891 cells with 744 mutations, a slight difference\n",
    "cells = df['cell_id'].drop_duplicates().tolist()\n",
    "sites = df['snp_id'].drop_duplicates().tolist()\n",
    "print(len(cells), len(sites))\n",
    "prefix = f'ov2295_{len(sites)}_{len(cells)}'\n",
    "print(prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a meta dataframe for the data\n",
    "df_site_meta = df[['snp_id', 'chrom', 'coord', 'ref', 'alt']].drop_duplicates()\n",
    "df_site_meta.sort_values(by='snp_id', inplace=True)\n",
    "df_site_meta['id'] = [f's{ix}' for ix in np.arange(df_site_meta.shape[0])]\n",
    "# df_site_meta.to_csv(f'./{prefix}_meta_site.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_meta = df[['cell_id', 'sample_id']].drop_duplicates().sort_values(by='cell_id')\n",
    "df_cell_meta['id'] = [ix for ix in np.arange(df_cell_meta.shape[0])]\n",
    "df_cell_meta = pd.merge(df_cell_meta, df_clones, on='cell_id')\n",
    "# df_cell_meta.to_csv(f'./{prefix}_meta_cell.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get af for each site\n",
    "def get_ml_gt(ref_counts, alt_counts):\n",
    "    a, b, c = phred_likelihood_with_fn_fp_flat(ref_counts, alt_counts)\n",
    "    min_phred = min(a, b, c) \n",
    "    if a == min_phred:\n",
    "        return 0\n",
    "    elif b == min_phred:\n",
    "        return 1\n",
    "    else:\n",
    "        return 2\n",
    "\n",
    "df['ml_gt'] = df.apply(lambda x: get_ml_gt(x['ref_counts'], x['alt_counts']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calcutate the phred likelihood for each site\n",
    "def get_phred_likelihood(ref_counts, alt_counts, ml_gt):\n",
    "    gts = ['0/0', '0/1', '1/1']\n",
    "    a, b, c = phred_likelihood_with_fn_fp_flat(ref_counts, alt_counts)\n",
    "    return f'{gts[ml_gt]}:{a},{b},{c}'\n",
    "df['phred'] = df.apply(lambda x: get_phred_likelihood(x['ref_counts'], x['alt_counts'], x['ml_gt']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def af(gts):\n",
    "    counts = [0, 0, 0]\n",
    "    for gt in gts:\n",
    "        counts[gt] += 1\n",
    "    counts = np.array(counts)\n",
    "    freqs = counts / sum(counts)\n",
    "    return np.round(freqs[0] + 0.5 * freqs[1], 2)\n",
    "\n",
    "allele_freqs = df.groupby('snp_id')['ml_gt'].agg(func=af).to_frame().rename(columns={'ml_gt': 'af'})\n",
    "# del df['af']\n",
    "# del df_site_meta['af']\n",
    "df = df.join(allele_freqs, on='snp_id')\n",
    "df_site_meta = df_site_meta.join(allele_freqs, on='snp_id')\n",
    "# df_site_meta.to_csv('./meta_site.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the posterior probability for each site\n",
    "df['genotype_prob'] = df.apply(lambda x: posterior_prob_with_fn_fp_flat_gt(x['ref_counts'], x['alt_counts'], prior_ref=x['af']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_to_scistree(genotype_matrix):\n",
    "    nsite, ncell = genotype_matrix.shape\n",
    "    output = f'ov2295_{nsite}_{ncell}.scistree'\n",
    "    with open(output, 'w') as out:\n",
    "        out.write(f'HAPLOID {nsite} {ncell}')\n",
    "        for i in range(ncell):\n",
    "            out.write(f' {i}')\n",
    "        out.write('\\n')\n",
    "        # for i in range(self.nsite):\n",
    "        # idx = 1\n",
    "        for i in range(nsite):\n",
    "            out.write(f's{i}')\n",
    "            # idx += 1\n",
    "            for j in range(ncell):\n",
    "                prob = genotype_matrix[i, j]\n",
    "                out.write(f' {prob:.5f}')\n",
    "            out.write('\\n')\n",
    "\n",
    "def write_to_huntress(matrix):\n",
    "    flags = [0, 1, 1, 3]\n",
    "    nsite, ncell = matrix.shape\n",
    "    output = f'ov2295_{nsite}_{ncell}.huntress'\n",
    "    with open(output, 'w') as out:\n",
    "        out.write(f'cellID/mutID')\n",
    "        for i in range(nsite):\n",
    "            out.write(f'\\ts{i}')\n",
    "        out.write('\\n')\n",
    "        for i in range(ncell):\n",
    "            out.write(f'c{i}')\n",
    "            for j in range(nsite):\n",
    "                val = flags[int(matrix[j, i])]\n",
    "                out.write(f'\\t{val}')\n",
    "            out.write('\\n')\n",
    "\n",
    "def write_to_cellphy(matrix):\n",
    "    nsite, ncell = matrix.shape\n",
    "    output = f'ov2295_{nsite}_{ncell}.vcf'\n",
    "    header = \\\n",
    "f'''##fileformat=VCFv4.3\n",
    "##fileDate={date.today().isoformat()}\n",
    "##source=ov2295\n",
    "##ncell={ncell}\n",
    "##nsite={nsite}\n",
    "##reference=NONE\n",
    "##contig=<ID=1>\n",
    "##phasing=NO\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Phread-scaled genotype likelihoods\">\n",
    "'''\n",
    "    with open(output, 'w') as out:\n",
    "        out.write(header)\n",
    "        out.write('#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT')\n",
    "        for i in range(ncell):\n",
    "            out.write(f'\\t{i}')\n",
    "        out.write('\\n')\n",
    "        for i in range(nsite):\n",
    "            chrom, coord, ref, alt, af, snp_id = df_site_meta.iloc[i][['chrom', 'coord', 'ref', 'alt', 'af', 'snp_id']]\n",
    "            out.write(f\"{chrom}\\t{coord}\\t{snp_id}\\t{ref}\\t{alt}\\t.\\tPASS\\tAF={af}\\tGT:PL\")\n",
    "            for j in range(ncell):\n",
    "                out.write(f'\\t{matrix[i, j]}')\n",
    "            out.write('\\n')                               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # formatted as scistree input\n",
    "# scistree_matrix_df = df.pivot(index='snp_id', columns='cell_id', values='genotype_prob')\n",
    "# scistree_matrix = scistree_matrix_df.to_numpy()\n",
    "# genotype_matrix_scistree = np.nan_to_num(scistree_matrix, nan=0.5)\n",
    "# write_to_scistree(genotype_matrix_scistree) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # formatted as huntress input\n",
    "# huntress_matrix_df = df.pivot(index='snp_id', columns='cell_id', values='ml_gt')\n",
    "# huntress_matrix = huntress_matrix_df.to_numpy()\n",
    "# huntress_matrix = np.nan_to_num(huntress_matrix, nan=3)\n",
    "# write_to_huntress(huntress_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# formatted as cellphy input\n",
    "cellphy_matrix_df = df.pivot(index='snp_id', columns='cell_id', values='phred') \n",
    "cellphy_matrix_df.fillna('./.:0,0,0', inplace=True)\n",
    "cellphy_matrix = cellphy_matrix_df.to_numpy()\n",
    "write_to_cellphy(cellphy_matrix)\n",
    "# cellphy_matrix_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run scistree \n",
    "# input_scis = f'{prefix}.scistree'\n",
    "# ! /usr/bin/time -o {prefix}.scistree.time /data/haotian/scistree/scistree_bin/Code/Scistree-v2.1.1.1/scistree -v -T 30 -e -o {prefix}.prob.muttree.gml $input_scis > {prefix}.scistree.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run huntress\n",
    "# input_huntress = f'{prefix}.huntress'\n",
    "# ! /usr/bin/time -o {prefix}.huntress.time python /home/haz19024/softwares/HUNTRESS/HUNTRESS.py --i $input_huntress --o {prefix}.SC --t 30 > {prefix}.HUNTRESS.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run cellphy\n",
    "# input_cellphy = f'{prefix}.vcf'\n",
    "# ! /usr/bin/time -o {prefix}.cellphy.time /home/haz19024/softwares/cellphy/cellphy.sh FAST -a -t 30 -r $input_cellphy > {prefix}.cellphy.log 2>&1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results (HGSOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOTICE: Running with system command <convert2annovar.pl  -includeinfo -allsample -withfreq -format vcf4 ov2295_14068_891.vcf > ov2295_14068_891.avinput>\n",
      "NOTICE: Finished reading 14079 lines from VCF file\n",
      "NOTICE: A total of 14068 locus in VCF file passed QC threshold, representing 14068 SNPs (5707 transitions and 8361 transversions) and 0 indels/substitutions\n",
      "NOTICE: Finished writing allele frequencies based on 12534588 SNP genotypes (5084937 transitions and 7449651 transversions) and 0 indels/substitutions for 891 samples\n",
      "\n",
      "NOTICE: Running with system command </home/haz19024/softwares/annovar/table_annovar.pl ov2295_14068_891.avinput /home/haz19024/softwares/annovar/humandb/humandb/ -buildver hg19 -outfile ov2295_14068_891 -remove -protocol refGene,cytoBand,exac03,avsnp147,dbnsfp30a -operation gx,r,f,f,f -nastring . -polish -xref /home/haz19024/softwares/annovar/example/gene_fullxref.txt -otherinfo>\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=gx protocol=refGene\n",
      "\n",
      "NOTICE: Running with system command <annotate_variation.pl -geneanno -buildver hg19 -dbtype refGene -outfile ov2295_14068_891.refGene -exonsort -nofirstcodondel ov2295_14068_891.avinput /home/haz19024/softwares/annovar/humandb/humandb/>\n",
      "NOTICE: Output files are written to ov2295_14068_891.refGene.variant_function, ov2295_14068_891.refGene.exonic_variant_function\n",
      "NOTICE: Reading gene annotation from /home/haz19024/softwares/annovar/humandb/humandb/hg19_refGene.txt ... Done with 78239 transcripts (including 18578 without coding sequence annotation) for 28293 unique genes\n",
      "NOTICE: Processing next batch with 14068 unique variants in 14068 input lines\n",
      "NOTICE: Reading FASTA sequences from /home/haz19024/softwares/annovar/humandb/humandb/hg19_refGeneMrna.fa ... Done with 604 sequences\n",
      "WARNING: A total of 465 sequences will be ignored due to lack of correct ORF annotation\n",
      "\n",
      "NOTICE: Running with system command <coding_change.pl  ov2295_14068_891.refGene.exonic_variant_function.orig /home/haz19024/softwares/annovar/humandb/humandb//hg19_refGene.txt /home/haz19024/softwares/annovar/humandb/humandb//hg19_refGeneMrna.fa -alltranscript -out ov2295_14068_891.refGene.fa -newevf ov2295_14068_891.refGene.exonic_variant_function>\n",
      "NOTICE: The xrefkey is provided in header as <pLi pRec pNull Gene_full_name Function_description Disease_description Tissue_specificity(Uniprot) Expression(egenetics) Expression(GNF/Atlas) P(HI) P(rec) RVIS RVIS_percentile GDI GDI-Phred>\n",
      "NOTICE: Finished reading 597255 cross references (each with 15 fields) from /home/haz19024/softwares/annovar/example/gene_fullxref.txt\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=r protocol=cytoBand\n",
      "\n",
      "NOTICE: Running with system command <annotate_variation.pl -regionanno -dbtype cytoBand -buildver hg19 -outfile ov2295_14068_891 ov2295_14068_891.avinput /home/haz19024/softwares/annovar/humandb/humandb/>\n",
      "NOTICE: Output file is written to ov2295_14068_891.hg19_cytoBand\n",
      "NOTICE: Reading annotation database /home/haz19024/softwares/annovar/humandb/humandb/hg19_cytoBand.txt ... Done with 862 regions\n",
      "NOTICE: Finished region-based annotation on 14068 genetic variants\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=f protocol=exac03\n",
      "NOTICE: Finished reading 8 column headers for '-dbtype exac03'\n",
      "\n",
      "NOTICE: Running system command <annotate_variation.pl -filter -dbtype exac03 -buildver hg19 -outfile ov2295_14068_891 ov2295_14068_891.avinput /home/haz19024/softwares/annovar/humandb/humandb/ -otherinfo>\n",
      "NOTICE: Output file with variants matching filtering criteria is written to ov2295_14068_891.hg19_exac03_dropped, and output file with other variants is written to ov2295_14068_891.hg19_exac03_filtered\n",
      "NOTICE: Processing next batch with 14068 unique variants in 14068 input lines\n",
      "NOTICE: Database index loaded. Total number of bins is 749886 and the number of bins to be scanned is 403\n",
      "NOTICE: Scanning filter database /home/haz19024/softwares/annovar/humandb/humandb/hg19_exac03.txt...Done\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=f protocol=avsnp147\n",
      "\n",
      "NOTICE: Running system command <annotate_variation.pl -filter -dbtype avsnp147 -buildver hg19 -outfile ov2295_14068_891 ov2295_14068_891.avinput /home/haz19024/softwares/annovar/humandb/humandb/>\n",
      "NOTICE: Output file with variants matching filtering criteria is written to ov2295_14068_891.hg19_avsnp147_dropped, and output file with other variants is written to ov2295_14068_891.hg19_avsnp147_filtered\n",
      "NOTICE: Processing next batch with 14068 unique variants in 14068 input lines\n",
      "NOTICE: Database index loaded. Total number of bins is 27868332 and the number of bins to be scanned is 13618\n",
      "NOTICE: Scanning filter database /home/haz19024/softwares/annovar/humandb/humandb/hg19_avsnp147.txt...Done\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=f protocol=dbnsfp30a\n",
      "NOTICE: Finished reading 34 column headers for '-dbtype dbnsfp30a'\n",
      "\n",
      "NOTICE: Running system command <annotate_variation.pl -filter -dbtype dbnsfp30a -buildver hg19 -outfile ov2295_14068_891 ov2295_14068_891.avinput /home/haz19024/softwares/annovar/humandb/humandb/ -otherinfo>\n",
      "NOTICE: Output file with variants matching filtering criteria is written to ov2295_14068_891.hg19_dbnsfp30a_dropped, and output file with other variants is written to ov2295_14068_891.hg19_dbnsfp30a_filtered\n",
      "NOTICE: Processing next batch with 14068 unique variants in 14068 input lines\n",
      "NOTICE: Database index loaded. Total number of bins is 548895 and the number of bins to be scanned is 300\n",
      "NOTICE: Scanning filter database /home/haz19024/softwares/annovar/humandb/humandb/hg19_dbnsfp30a.txt...Done\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Multianno output file is written to ov2295_14068_891.hg19_multianno.txt\n",
      "NOTICE: Reading from ov2295_14068_891.hg19_multianno.txt\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: VCF output is written to ov2295_14068_891.hg19_multianno.vcf\n"
     ]
    }
   ],
   "source": [
    "# annovar annotation\n",
    "humandb = '/home/haz19024/softwares/annovar/humandb/humandb/'\n",
    "xref = '/home/haz19024/softwares/annovar/example/gene_fullxref.txt'\n",
    "!perl /home/haz19024/softwares/annovar/table_annovar.pl {prefix}.vcf $humandb -buildver hg19 -out {prefix} -remove -protocol refGene,cytoBand,exac03,avsnp147,dbnsfp30a -operation gx,r,f,f,f -nastring . -vcfinput -polish -xref $xref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/haz19024/.local/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3460: DtypeWarning: Columns (0,72) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>Otherinfo894</th>\n",
       "      <th>Otherinfo895</th>\n",
       "      <th>Otherinfo896</th>\n",
       "      <th>Otherinfo897</th>\n",
       "      <th>Otherinfo898</th>\n",
       "      <th>Otherinfo899</th>\n",
       "      <th>Otherinfo900</th>\n",
       "      <th>Otherinfo901</th>\n",
       "      <th>Otherinfo902</th>\n",
       "      <th>Otherinfo903</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10:100129846:C:A</th>\n",
       "      <td>10</td>\n",
       "      <td>100129846</td>\n",
       "      <td>100129846</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOXL4;PYROXD2</td>\n",
       "      <td>dist=101895;dist=13479</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:20,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100185101:A:G</th>\n",
       "      <td>10</td>\n",
       "      <td>100185101</td>\n",
       "      <td>100185101</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:40,7,1</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:40,7,1</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100209326:T:C</th>\n",
       "      <td>10</td>\n",
       "      <td>100209326</td>\n",
       "      <td>100209326</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>LOC101927278</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100283942:C:T</th>\n",
       "      <td>10</td>\n",
       "      <td>100283942</td>\n",
       "      <td>100283942</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>intronic</td>\n",
       "      <td>HPSE2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:10086005:G:T</th>\n",
       "      <td>10</td>\n",
       "      <td>10086005</td>\n",
       "      <td>10086005</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC101928272;LINC02670</td>\n",
       "      <td>dist=748449;dist=14680</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:98278474:C:A</th>\n",
       "      <td>X</td>\n",
       "      <td>98278474</td>\n",
       "      <td>98278474</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>DIAPH2;PCDH19</td>\n",
       "      <td>dist=1418479;dist=1268168</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:98452192:T:C</th>\n",
       "      <td>X</td>\n",
       "      <td>98452192</td>\n",
       "      <td>98452192</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>DIAPH2;PCDH19</td>\n",
       "      <td>dist=1592197;dist=1094450</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:98700894:C:T</th>\n",
       "      <td>X</td>\n",
       "      <td>98700894</td>\n",
       "      <td>98700894</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>DIAPH2;PCDH19</td>\n",
       "      <td>dist=1840899;dist=845748</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:20,3,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:99136386:C:T</th>\n",
       "      <td>X</td>\n",
       "      <td>99136386</td>\n",
       "      <td>99136386</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NONE;PCDH19</td>\n",
       "      <td>dist=NONE;dist=410256</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:20,3,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:99571629:G:T</th>\n",
       "      <td>X</td>\n",
       "      <td>99571629</td>\n",
       "      <td>99571629</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>intronic</td>\n",
       "      <td>PCDH19</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14068 rows Ã— 972 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Chr      Start        End Ref Alt    Func.refGene  \\\n",
       "variant_id                                                           \n",
       "10:100129846:C:A  10  100129846  100129846   C   A      intergenic   \n",
       "10:100185101:A:G  10  100185101  100185101   A   G        intronic   \n",
       "10:100209326:T:C  10  100209326  100209326   T   C  ncRNA_intronic   \n",
       "10:100283942:C:T  10  100283942  100283942   C   T        intronic   \n",
       "10:10086005:G:T   10   10086005   10086005   G   T      intergenic   \n",
       "...               ..        ...        ...  ..  ..             ...   \n",
       "X:98278474:C:A     X   98278474   98278474   C   A      intergenic   \n",
       "X:98452192:T:C     X   98452192   98452192   T   C      intergenic   \n",
       "X:98700894:C:T     X   98700894   98700894   C   T      intergenic   \n",
       "X:99136386:C:T     X   99136386   99136386   C   T      intergenic   \n",
       "X:99571629:G:T     X   99571629   99571629   G   T        intronic   \n",
       "\n",
       "                            Gene.refGene         GeneDetail.refGene  \\\n",
       "variant_id                                                            \n",
       "10:100129846:C:A           LOXL4;PYROXD2     dist=101895;dist=13479   \n",
       "10:100185101:A:G                    HPS1                          .   \n",
       "10:100209326:T:C            LOC101927278                          .   \n",
       "10:100283942:C:T                   HPSE2                          .   \n",
       "10:10086005:G:T   LOC101928272;LINC02670     dist=748449;dist=14680   \n",
       "...                                  ...                        ...   \n",
       "X:98278474:C:A             DIAPH2;PCDH19  dist=1418479;dist=1268168   \n",
       "X:98452192:T:C             DIAPH2;PCDH19  dist=1592197;dist=1094450   \n",
       "X:98700894:C:T             DIAPH2;PCDH19   dist=1840899;dist=845748   \n",
       "X:99136386:C:T               NONE;PCDH19      dist=NONE;dist=410256   \n",
       "X:99571629:G:T                    PCDH19                          .   \n",
       "\n",
       "                 ExonicFunc.refGene AAChange.refGene  ... Otherinfo894  \\\n",
       "variant_id                                            ...                \n",
       "10:100129846:C:A                  .                .  ...    ./.:0,0,0   \n",
       "10:100185101:A:G                  .                .  ...    ./.:0,0,0   \n",
       "10:100209326:T:C                  .                .  ...    ./.:0,0,0   \n",
       "10:100283942:C:T                  .                .  ...   0/0:0,4,13   \n",
       "10:10086005:G:T                   .                .  ...    ./.:0,0,0   \n",
       "...                             ...              ...  ...          ...   \n",
       "X:98278474:C:A                    .                .  ...    ./.:0,0,0   \n",
       "X:98452192:T:C                    .                .  ...    ./.:0,0,0   \n",
       "X:98700894:C:T                    .                .  ...    ./.:0,0,0   \n",
       "X:99136386:C:T                    .                .  ...    ./.:0,0,0   \n",
       "X:99571629:G:T                    .                .  ...    ./.:0,0,0   \n",
       "\n",
       "                 Otherinfo895 Otherinfo896 Otherinfo897 Otherinfo898  \\\n",
       "variant_id                                                             \n",
       "10:100129846:C:A    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100185101:A:G   1/1:40,7,1    0/0:0,2,6    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100209326:T:C    0/0:0,2,6    ./.:0,0,0    0/0:0,2,6    ./.:0,0,0   \n",
       "10:100283942:C:T    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:10086005:G:T     0/0:0,2,6    0/0:0,2,6    ./.:0,0,0    ./.:0,0,0   \n",
       "...                       ...          ...          ...          ...   \n",
       "X:98278474:C:A      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "X:98452192:T:C      ./.:0,0,0    ./.:0,0,0    0/0:0,2,6    ./.:0,0,0   \n",
       "X:98700894:C:T      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "X:99136386:C:T     1/1:20,3,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "X:99571629:G:T      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "\n",
       "                 Otherinfo899 Otherinfo900 Otherinfo901 Otherinfo902  \\\n",
       "variant_id                                                             \n",
       "10:100129846:C:A    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100185101:A:G   1/1:40,7,1    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100209326:T:C    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100283942:C:T    ./.:0,0,0    ./.:0,0,0    0/0:0,2,6    ./.:0,0,0   \n",
       "10:10086005:G:T     ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "...                       ...          ...          ...          ...   \n",
       "X:98278474:C:A      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "X:98452192:T:C      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "X:98700894:C:T      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   1/1:20,3,0   \n",
       "X:99136386:C:T      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "X:99571629:G:T      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    0/0:0,2,6   \n",
       "\n",
       "                 Otherinfo903  \n",
       "variant_id                     \n",
       "10:100129846:C:A   1/1:20,3,0  \n",
       "10:100185101:A:G   0/0:0,4,13  \n",
       "10:100209326:T:C    ./.:0,0,0  \n",
       "10:100283942:C:T    0/0:0,2,6  \n",
       "10:10086005:G:T     ./.:0,0,0  \n",
       "...                       ...  \n",
       "X:98278474:C:A      ./.:0,0,0  \n",
       "X:98452192:T:C      ./.:0,0,0  \n",
       "X:98700894:C:T      ./.:0,0,0  \n",
       "X:99136386:C:T      ./.:0,0,0  \n",
       "X:99571629:G:T      ./.:0,0,0  \n",
       "\n",
       "[14068 rows x 972 columns]"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene annotation\n",
    "df_gene_annotation = pd.read_table(f'./{prefix}.hg19_multianno.txt')\n",
    "df_gene_annotation['variant_id'] = df_gene_annotation['Chr'].astype(str) + ':' + df_gene_annotation['Start'].astype(str) + ':' + df_gene_annotation['Ref'] + ':' + df_gene_annotation['Alt']\n",
    "df_gene_annotation.set_index('variant_id', inplace=True)\n",
    "df_gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ABCA13', 'ABCA9', 'ACBD3', 'ADAM12', 'ADGB', 'ADGRF3', 'ADRA2B', 'AGRN', 'AKAP9', 'ALPK1', 'APOB', 'APOL1', 'APOL3', 'ARAP3', 'ARID1A', 'ARNT', 'ASB15', 'ASIC2', 'ASL', 'ATAD2B', 'ATXN2L', 'BIRC6', 'BTBD11', 'C2orf42', 'CACNA1G', 'CASR', 'CBL', 'CCDC180', 'CCDC77', 'CDC123', 'CDH23', 'CHD2', 'CLDN17', 'CRISPLD1', 'CSMD3', 'CUX2', 'DEPDC5', 'DGKQ', 'DHRS13', 'DNAH17', 'DNAH2', 'DNAH9', 'DNHD1', 'DUOX1', 'ECM2', 'EDAR', 'EFNA3', 'EFS', 'EGR2', 'ELAPOR1', 'EMILIN2', 'EML2', 'ERN2', 'ESPNL', 'ESRRG', 'FAM124A', 'FAM170B', 'FAM83H', 'FBXO32', 'FBXW5', 'FCRL3', 'FNDC1', 'FOXP2', 'GAB4', 'GANC', 'GDF7', 'GEM', 'GHR', 'GLG1', 'GNAT1', 'GRIK1', 'GRIK3', 'GRIK5', 'GRIN3A', 'GSTM3', 'GUCA1ANB', 'HDGFL1', 'HOOK3', 'HTR1D', 'HTRA4', 'INSL4', 'INTS5', 'IRF4', 'JPH1', 'KANK1', 'KANSL1', 'KANSL1L', 'KDM2B', 'KIF1A', 'KIZ', 'KLHL33', 'LAMA2', 'LPCAT1', 'LRRC31', 'LRRC3C', 'LTN1', 'LY6D', 'MAGEB10', 'MAP1S', 'MARCOL', 'MCOLN2', 'MDN1', 'MED12L', 'METTL18', 'MMAA', 'MRPL21', 'MTREX', 'MYH9', 'NAT10', 'NCOA6', 'NEB', 'NEURL4', 'NID1', 'NOA1', 'NPM2', 'NPTX1', 'NPTXR', 'NRCAM', 'NSD1', 'OR2L13', 'PACS1', 'PARD3B', 'PARP11', 'PATE2', 'PBDC1', 'PBK', 'PCDH17', 'PCYT1B', 'PDS5B', 'PEG3', 'PHF12', 'POLE', 'PPP1R3A', 'PREX2', 'PROCR', 'PROSER2', 'PROX2', 'PSKH2', 'PTPN3', 'PTPRA', 'RARRES1', 'RASAL2', 'RINT1', 'RNF40', 'RPGRIP1', 'RPRD1A', 'RYR1', 'RYR2', 'SARDH', 'SEC22C', 'SGO2', 'SHISAL1', 'SLC19A3', 'SLC6A12', 'SMG1', 'SORBS2', 'SPRYD3', 'SRRT', 'STK10', 'SUGCT', 'SUSD3', 'SYNE2', 'SYNM', 'TACR3', 'TATDN2', 'TCL1A', 'TFDP2', 'THEM6', 'TIGD5', 'TLE2', 'TLN2', 'TMEM204', 'TMEM213', 'TOP2A', 'TP53', 'TRAPPC8', 'TSPYL5', 'TTLL7', 'TYSND1', 'UPK1A', 'USP34', 'VCL', 'WDFY3', 'WDR7', 'XRCC2', 'XXYLT1', 'ZC3H13', 'ZG16B', 'ZHX1', 'ZIC1', 'ZNF404', 'ZNF469', 'ZNF704', 'ZNF710', 'ZNF839']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>Otherinfo894</th>\n",
       "      <th>Otherinfo895</th>\n",
       "      <th>Otherinfo896</th>\n",
       "      <th>Otherinfo897</th>\n",
       "      <th>Otherinfo898</th>\n",
       "      <th>Otherinfo899</th>\n",
       "      <th>Otherinfo900</th>\n",
       "      <th>Otherinfo901</th>\n",
       "      <th>Otherinfo902</th>\n",
       "      <th>Otherinfo903</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>10</td>\n",
       "      <td>11912073</td>\n",
       "      <td>11912073</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>PROSER2</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>PROSER2:NM_153256:exon4:c.G976A:p.A326T</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>10</td>\n",
       "      <td>12257806</td>\n",
       "      <td>12257806</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>CDC123</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>CDC123:NM_006023:exon5:c.T305C:p.F102S</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:20,3,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>10</td>\n",
       "      <td>127730736</td>\n",
       "      <td>127730736</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>ADAM12</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ADAM12:NM_001288974:exon19:c.G2179A:p.E727K,AD...</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>10</td>\n",
       "      <td>50339712</td>\n",
       "      <td>50339712</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>FAM170B</td>\n",
       "      <td>.</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>FAM170B:NM_001164484:exon2:c.A798G:p.E266E</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0,9,27</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>10</td>\n",
       "      <td>64573400</td>\n",
       "      <td>64573400</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>EGR2</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>EGR2:NM_000399:exon2:c.C998T:p.T333M,EGR2:NM_0...</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1:20,3,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13610</th>\n",
       "      <td>9</td>\n",
       "      <td>95267903</td>\n",
       "      <td>95267903</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>ECM2</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ECM2:NM_001197295:exon7:c.C1310G:p.A437G,ECM2:...</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13613</th>\n",
       "      <td>9</td>\n",
       "      <td>95838138</td>\n",
       "      <td>95838138</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SUSD3</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>SUSD3:NM_001287006:exon2:c.G161A:p.G54E,SUSD3:...</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13843</th>\n",
       "      <td>X</td>\n",
       "      <td>24593304</td>\n",
       "      <td>24593304</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>PCYT1B</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>PCYT1B:NM_001163264:exon7:c.G786C:p.E262D,PCYT...</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1:40,7,1</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:40,7,1</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13855</th>\n",
       "      <td>X</td>\n",
       "      <td>27839987</td>\n",
       "      <td>27839987</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>MAGEB10</td>\n",
       "      <td>.</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>MAGEB10:NM_182506:exon3:c.T564C:p.C188C</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13990</th>\n",
       "      <td>X</td>\n",
       "      <td>75393475</td>\n",
       "      <td>75393475</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>PBDC1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>PBDC1:NM_001300888:exon2:c.A86G:p.Y29C,PBDC1:N...</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows Ã— 972 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr      Start        End Ref Alt Func.refGene Gene.refGene  \\\n",
       "147    10   11912073   11912073   G   A       exonic      PROSER2   \n",
       "169    10   12257806   12257806   T   C       exonic       CDC123   \n",
       "185    10  127730736  127730736   C   T       exonic       ADAM12   \n",
       "417    10   50339712   50339712   T   C       exonic      FAM170B   \n",
       "512    10   64573400   64573400   G   A       exonic         EGR2   \n",
       "...    ..        ...        ...  ..  ..          ...          ...   \n",
       "13610   9   95267903   95267903   G   C       exonic         ECM2   \n",
       "13613   9   95838138   95838138   G   A       exonic        SUSD3   \n",
       "13843   X   24593304   24593304   C   G       exonic       PCYT1B   \n",
       "13855   X   27839987   27839987   T   C       exonic      MAGEB10   \n",
       "13990   X   75393475   75393475   A   G       exonic        PBDC1   \n",
       "\n",
       "      GeneDetail.refGene ExonicFunc.refGene  \\\n",
       "147                    .  nonsynonymous SNV   \n",
       "169                    .  nonsynonymous SNV   \n",
       "185                    .  nonsynonymous SNV   \n",
       "417                    .     synonymous SNV   \n",
       "512                    .  nonsynonymous SNV   \n",
       "...                  ...                ...   \n",
       "13610                  .  nonsynonymous SNV   \n",
       "13613                  .  nonsynonymous SNV   \n",
       "13843                  .  nonsynonymous SNV   \n",
       "13855                  .     synonymous SNV   \n",
       "13990                  .  nonsynonymous SNV   \n",
       "\n",
       "                                        AAChange.refGene  ... Otherinfo894  \\\n",
       "147              PROSER2:NM_153256:exon4:c.G976A:p.A326T  ...    ./.:0,0,0   \n",
       "169               CDC123:NM_006023:exon5:c.T305C:p.F102S  ...    ./.:0,0,0   \n",
       "185    ADAM12:NM_001288974:exon19:c.G2179A:p.E727K,AD...  ...    ./.:0,0,0   \n",
       "417           FAM170B:NM_001164484:exon2:c.A798G:p.E266E  ...   0/0:0,9,27   \n",
       "512    EGR2:NM_000399:exon2:c.C998T:p.T333M,EGR2:NM_0...  ...   1/1:20,3,0   \n",
       "...                                                  ...  ...          ...   \n",
       "13610  ECM2:NM_001197295:exon7:c.C1310G:p.A437G,ECM2:...  ...    ./.:0,0,0   \n",
       "13613  SUSD3:NM_001287006:exon2:c.G161A:p.G54E,SUSD3:...  ...    ./.:0,0,0   \n",
       "13843  PCYT1B:NM_001163264:exon7:c.G786C:p.E262D,PCYT...  ...   1/1:40,7,1   \n",
       "13855            MAGEB10:NM_182506:exon3:c.T564C:p.C188C  ...    0/0:0,2,6   \n",
       "13990  PBDC1:NM_001300888:exon2:c.A86G:p.Y29C,PBDC1:N...  ...    ./.:0,0,0   \n",
       "\n",
       "      Otherinfo895 Otherinfo896 Otherinfo897 Otherinfo898 Otherinfo899  \\\n",
       "147      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "169      ./.:0,0,0    ./.:0,0,0   1/1:20,3,0    ./.:0,0,0   0/0:0,4,13   \n",
       "185      ./.:0,0,0    0/0:0,2,6   0/0:0,4,13    ./.:0,0,0    ./.:0,0,0   \n",
       "417      ./.:0,0,0    0/0:0,2,6    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "512      ./.:0,0,0    ./.:0,0,0   0/0:0,4,13    ./.:0,0,0   0/0:0,4,13   \n",
       "...            ...          ...          ...          ...          ...   \n",
       "13610    0/0:0,2,6    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "13613    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "13843    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "13855    ./.:0,0,0    ./.:0,0,0    0/0:0,2,6    ./.:0,0,0    ./.:0,0,0   \n",
       "13990    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "\n",
       "      Otherinfo900 Otherinfo901 Otherinfo902 Otherinfo903  \n",
       "147      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "169      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "185      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "417      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "512      ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "...            ...          ...          ...          ...  \n",
       "13610    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "13613    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "13843    ./.:0,0,0    ./.:0,0,0   1/1:40,7,1    ./.:0,0,0  \n",
       "13855    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "13990    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0  \n",
       "\n",
       "[204 rows x 972 columns]"
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find cancer-driver genes \n",
    "df_gene_exonic = df_gene_annotation[df_gene_annotation['Func.refGene'] == 'exonic']\n",
    "print(sorted(df_gene_exonic['Gene.refGene'].unique()))\n",
    "df_gene_exonic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>Gene.refGene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3839</th>\n",
       "      <td>17:7578265:A:G</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5179</th>\n",
       "      <td>1:23520184:G:A</td>\n",
       "      <td>HTR1D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11388</th>\n",
       "      <td>7:114329967:G:A</td>\n",
       "      <td>FOXP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11789</th>\n",
       "      <td>7:40900013:T:C</td>\n",
       "      <td>SUGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12200</th>\n",
       "      <td>8:124266354:T:A</td>\n",
       "      <td>ZHX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13443</th>\n",
       "      <td>9:5231710:C:T</td>\n",
       "      <td>INSL4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            variant_id Gene.refGene\n",
       "3839    17:7578265:A:G         TP53\n",
       "5179    1:23520184:G:A        HTR1D\n",
       "11388  7:114329967:G:A        FOXP2\n",
       "11789   7:40900013:T:C        SUGCT\n",
       "12200  8:124266354:T:A         ZHX1\n",
       "13443    9:5231710:C:T        INSL4"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# original paper used the following genes\n",
    "marker_genes = ['TP53', 'FOXP2', 'ZHX1', 'HTR1D', 'INSL4', 'C7orf10', 'SUGCT']\n",
    "df_marker_genes = df_gene_annotation[df_gene_annotation['Gene.refGene'].isin(marker_genes)]\n",
    "df_marker_genes = df_marker_genes[df_marker_genes['Func.refGene'] == 'exonic']\n",
    "df_marker_genes['variant_id'] = df_marker_genes.apply(func=lambda x: str(x['Chr'])+':'+str(x['Start'])+ ':' + x['Ref'] + ':' + x['Alt'], axis=1)\n",
    "df_marker_genes = df_marker_genes[['variant_id', 'Gene.refGene']]\n",
    "df_marker_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import dollo\n",
    "import pickle\n",
    "import popgen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>node</th>\n",
       "      <th>loss</th>\n",
       "      <th>origin</th>\n",
       "      <th>presence</th>\n",
       "      <th>ml_origin</th>\n",
       "      <th>ml_presence</th>\n",
       "      <th>ml_loss</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>id</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10:100129846:C:A</th>\n",
       "      <th>9</th>\n",
       "      <td>10:100129846:C:A</td>\n",
       "      <td>9</td>\n",
       "      <td>3.985202e-41</td>\n",
       "      <td>0.993210</td>\n",
       "      <td>0.006790</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10:100129846:C:A</td>\n",
       "      <td>10</td>\n",
       "      <td>100129846</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>s0</td>\n",
       "      <td>0.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100185101:A:G</th>\n",
       "      <th>26</th>\n",
       "      <td>10:100185101:A:G</td>\n",
       "      <td>9</td>\n",
       "      <td>3.896468e-67</td>\n",
       "      <td>0.993189</td>\n",
       "      <td>0.006811</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10:100185101:A:G</td>\n",
       "      <td>10</td>\n",
       "      <td>100185101</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>s1</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100209326:T:C</th>\n",
       "      <th>34</th>\n",
       "      <td>10:100209326:T:C</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10:100209326:T:C</td>\n",
       "      <td>10</td>\n",
       "      <td>100209326</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>s2</td>\n",
       "      <td>0.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100283942:C:T</th>\n",
       "      <th>60</th>\n",
       "      <td>10:100283942:C:T</td>\n",
       "      <td>9</td>\n",
       "      <td>5.161910e-86</td>\n",
       "      <td>0.993218</td>\n",
       "      <td>0.006782</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10:100283942:C:T</td>\n",
       "      <td>10</td>\n",
       "      <td>100283942</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>s3</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:10086005:G:T</th>\n",
       "      <th>70</th>\n",
       "      <td>10:10086005:G:T</td>\n",
       "      <td>2</td>\n",
       "      <td>5.890891e-61</td>\n",
       "      <td>0.993219</td>\n",
       "      <td>0.006781</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10:10086005:G:T</td>\n",
       "      <td>10</td>\n",
       "      <td>10086005</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>s4</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:98278474:C:A</th>\n",
       "      <th>239083</th>\n",
       "      <td>X:98278474:C:A</td>\n",
       "      <td>12</td>\n",
       "      <td>1.019073e-35</td>\n",
       "      <td>0.791074</td>\n",
       "      <td>0.005422</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>X:98278474:C:A</td>\n",
       "      <td>X</td>\n",
       "      <td>98278474</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>s14063</td>\n",
       "      <td>0.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:98452192:T:C</th>\n",
       "      <th>239093</th>\n",
       "      <td>X:98452192:T:C</td>\n",
       "      <td>5</td>\n",
       "      <td>4.754889e-44</td>\n",
       "      <td>0.993219</td>\n",
       "      <td>0.006781</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>X:98452192:T:C</td>\n",
       "      <td>X</td>\n",
       "      <td>98452192</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>s14064</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:98700894:C:T</th>\n",
       "      <th>239105</th>\n",
       "      <td>X:98700894:C:T</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>X:98700894:C:T</td>\n",
       "      <td>X</td>\n",
       "      <td>98700894</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>s14065</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:99136386:C:T</th>\n",
       "      <th>239131</th>\n",
       "      <td>X:99136386:C:T</td>\n",
       "      <td>9</td>\n",
       "      <td>4.789471e-86</td>\n",
       "      <td>0.993212</td>\n",
       "      <td>0.006788</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>X:99136386:C:T</td>\n",
       "      <td>X</td>\n",
       "      <td>99136386</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>s14066</td>\n",
       "      <td>0.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X:99571629:G:T</th>\n",
       "      <th>239151</th>\n",
       "      <td>X:99571629:G:T</td>\n",
       "      <td>12</td>\n",
       "      <td>4.783274e-116</td>\n",
       "      <td>0.993218</td>\n",
       "      <td>0.006782</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>X:99571629:G:T</td>\n",
       "      <td>X</td>\n",
       "      <td>99571629</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>s14067</td>\n",
       "      <td>0.65</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14068 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               variant_id  node           loss    origin  \\\n",
       "variant_id                                                                 \n",
       "10:100129846:C:A 9       10:100129846:C:A     9   3.985202e-41  0.993210   \n",
       "10:100185101:A:G 26      10:100185101:A:G     9   3.896468e-67  0.993189   \n",
       "10:100209326:T:C 34      10:100209326:T:C     0   0.000000e+00  1.000000   \n",
       "10:100283942:C:T 60      10:100283942:C:T     9   5.161910e-86  0.993218   \n",
       "10:10086005:G:T  70       10:10086005:G:T     2   5.890891e-61  0.993219   \n",
       "...                                   ...   ...            ...       ...   \n",
       "X:98278474:C:A   239083    X:98278474:C:A    12   1.019073e-35  0.791074   \n",
       "X:98452192:T:C   239093    X:98452192:T:C     5   4.754889e-44  0.993219   \n",
       "X:98700894:C:T   239105    X:98700894:C:T     0   0.000000e+00  1.000000   \n",
       "X:99136386:C:T   239131    X:99136386:C:T     9   4.789471e-86  0.993212   \n",
       "X:99571629:G:T   239151    X:99571629:G:T    12  4.783274e-116  0.993218   \n",
       "\n",
       "                         presence  ml_origin  ml_presence  ml_loss  \\\n",
       "variant_id                                                           \n",
       "10:100129846:C:A 9       0.006790          1            1        0   \n",
       "10:100185101:A:G 26      0.006811          1            1        0   \n",
       "10:100209326:T:C 34      0.000000          1            1        0   \n",
       "10:100283942:C:T 60      0.006782          1            1        0   \n",
       "10:10086005:G:T  70      0.006781          1            1        0   \n",
       "...                           ...        ...          ...      ...   \n",
       "X:98278474:C:A   239083  0.005422          1            1        0   \n",
       "X:98452192:T:C   239093  0.006781          1            1        0   \n",
       "X:98700894:C:T   239105  0.000000          1            1        0   \n",
       "X:99136386:C:T   239131  0.006788          1            1        0   \n",
       "X:99571629:G:T   239151  0.006782          1            1        0   \n",
       "\n",
       "                                   snp_id chrom      coord ref alt      id  \\\n",
       "variant_id                                                                   \n",
       "10:100129846:C:A 9       10:100129846:C:A    10  100129846   C   A      s0   \n",
       "10:100185101:A:G 26      10:100185101:A:G    10  100185101   A   G      s1   \n",
       "10:100209326:T:C 34      10:100209326:T:C    10  100209326   T   C      s2   \n",
       "10:100283942:C:T 60      10:100283942:C:T    10  100283942   C   T      s3   \n",
       "10:10086005:G:T  70       10:10086005:G:T    10   10086005   G   T      s4   \n",
       "...                                   ...   ...        ...  ..  ..     ...   \n",
       "X:98278474:C:A   239083    X:98278474:C:A     X   98278474   C   A  s14063   \n",
       "X:98452192:T:C   239093    X:98452192:T:C     X   98452192   T   C  s14064   \n",
       "X:98700894:C:T   239105    X:98700894:C:T     X   98700894   C   T  s14065   \n",
       "X:99136386:C:T   239131    X:99136386:C:T     X   99136386   C   T  s14066   \n",
       "X:99571629:G:T   239151    X:99571629:G:T     X   99571629   G   T  s14067   \n",
       "\n",
       "                           af  \n",
       "variant_id                     \n",
       "10:100129846:C:A 9       0.88  \n",
       "10:100185101:A:G 26      0.83  \n",
       "10:100209326:T:C 34      0.38  \n",
       "10:100283942:C:T 60      0.86  \n",
       "10:10086005:G:T  70      0.92  \n",
       "...                       ...  \n",
       "X:98278474:C:A   239083  0.89  \n",
       "X:98452192:T:C   239093  0.80  \n",
       "X:98700894:C:T   239105  0.00  \n",
       "X:99136386:C:T   239131  0.73  \n",
       "X:99571629:G:T   239151  0.65  \n",
       "\n",
       "[14068 rows x 15 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('../ov2295_tree.pickle', 'rb') as f:\n",
    "    tree = pickle.load(f)\n",
    "df_nodes = pd.read_csv('../ov2295_nodes.csv')\n",
    "df_mutations = pd.merge(df_nodes, df_site_meta, left_on='variant_id', right_on='snp_id')\n",
    "df_mutations = df_mutations.groupby(['variant_id']).apply(lambda x: x[x['ml_origin']==1])\n",
    "df_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_marker_genes.set_index('variant_id', inplace=True)\n",
    "df_mutations.set_index('variant_id', inplace=True)\n",
    "df_marker_genes = df_marker_genes.join(df_mutations, on='variant_id')\n",
    "df_marker_genes.set_index('id', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "# read outputs\n",
    "def get_scistree_genotype(prefix):\n",
    "    geno_file = f'{prefix}.scistree.genos.imp'\n",
    "    genotypes = []\n",
    "    with open(geno_file, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('Site'):\n",
    "                line = line.strip()\n",
    "                genos = line.split('\\t')[1].split()\n",
    "                genos = list(map(int, genos))\n",
    "                genotypes.append(genos)\n",
    "    return np.array(genotypes)\n",
    "\n",
    "def get_huntress_genotype(prefix):\n",
    "    geno_file = f'{prefix}.SC.CFMatrix'\n",
    "    genotypes = pd.read_table(geno_file, index_col=0).values.T\n",
    "    return genotypes\n",
    "\n",
    "def read_mutation_list(file):\n",
    "    edges = {}\n",
    "    with open(file, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            items = line.strip().split()\n",
    "            if int(items[1]) > 0:\n",
    "                eid, num_muts, muts = items\n",
    "                edges[eid] = [int(_[1:]) for _ in muts.split(',')]\n",
    "            else:\n",
    "                eid = items[0]\n",
    "                edges[eid] = []\n",
    "    return edges\n",
    "\n",
    "\n",
    "def read_mutation_tree(file):\n",
    "    with open(file, 'r') as f:\n",
    "        line = f.readline()\n",
    "        line = line.strip()\n",
    "    return popgen.utils.from_newick(line)\n",
    "\n",
    "def get_cellphy_genotype(prefix):\n",
    "    raxml_tree = f'{prefix}.vcf.Mapped.raxml.mutationMapTree'\n",
    "    raxml_list = f'{prefix}.vcf.Mapped.raxml.mutationMapList'\n",
    "    num_snp, num_cell = len(sites), len(cells)\n",
    "    genotype = np.zeros([num_snp, num_cell], dtype=int)\n",
    "    edges = read_mutation_list(raxml_list)\n",
    "    tree = read_mutation_tree(raxml_tree)\n",
    "    generator = popgen.utils.TraversalGenerator(order='post')\n",
    "    for node in generator(tree):\n",
    "        if not node.is_root():\n",
    "            eid = re.search(r'\\[(\\d+)\\]', node.branch).group(1)\n",
    "            muts = edges[eid]\n",
    "            # muts = [mut-1 for mut in muts] # to 0-indexed\n",
    "            leaves = [int(leaf.name) for leaf in node.get_leaves()] # to 0-indexed\n",
    "            genotype[np.ix_(muts, leaves)] = 1\n",
    "    return genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get scistree genotype\n",
    "scistree_genotype = get_scistree_genotype(prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_covered(clade1, clade2):\n",
    "    diff = clade1 - clade2\n",
    "    if 0 in diff and 1 in diff and -1 in diff:\n",
    "        return 0 # no relationship\n",
    "    if 1 in diff:\n",
    "        return 1 # clade1(snp1) happens before clade2(snp2)\n",
    "    else:\n",
    "        return -1 # clade2(snp2) happens before clade1(snp1)\n",
    "    \n",
    "def get_ancestor_descendant_pairs_4_given_genes(geno, df_marker_genes):\n",
    "    genes = df_marker_genes.index.tolist()\n",
    "    num_snp, _ = geno.shape\n",
    "    mutations = {df_marker_genes.loc[gene, 'Gene.refGene']: set() for gene in genes}\n",
    "    for i in range(len(genes)):\n",
    "        for j in range(len(genes)):\n",
    "            if i == j:\n",
    "                continue\n",
    "            gene1, gene2 = genes[i], genes[j]\n",
    "            gene1_idx = int(gene1.split('s')[1])\n",
    "            gene2_idx = int(gene2.split('s')[1])\n",
    "            code = is_covered(geno[gene1_idx], geno[gene2_idx])\n",
    "            if code == 1:\n",
    "                mutations[df_marker_genes.loc[gene1, 'Gene.refGene']].add(df_marker_genes.loc[gene2, 'Gene.refGene'])\n",
    "            elif code == -1:\n",
    "                mutations[df_marker_genes.loc[gene2, 'Gene.refGene']].add(df_marker_genes.loc[gene1, 'Gene.refGene'])\n",
    "    return mutations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TP53': {'FOXP2', 'HTR1D', 'INSL4', 'SUGCT', 'ZHX1'},\n",
       " 'HTR1D': {'INSL4'},\n",
       " 'FOXP2': {'HTR1D', 'INSL4', 'SUGCT', 'TP53', 'ZHX1'},\n",
       " 'SUGCT': {'FOXP2', 'HTR1D', 'INSL4', 'TP53', 'ZHX1'},\n",
       " 'ZHX1': set(),\n",
       " 'INSL4': set()}"
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# interested mutations order, Ancestral mutations with significant impact were found in TP53 (584T>C), FOXP2, and SUGCT\n",
    "# Clade specific mutations were found in ZHX1, HTR1D, INSL4\n",
    "scistree_mutations = get_ancestor_descendant_pairs_4_given_genes(scistree_genotype, df_marker_genes)\n",
    "scistree_mutations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the mutations in the tree\n",
    "marker_gene_clades = {}\n",
    "for gene in marker_genes:\n",
    "    ix = df_marker_genes[df_marker_genes['Gene.refGene'] == gene].index.tolist()\n",
    "    if ix:\n",
    "        genotype = scistree_genotype[int(ix[0][1:])]\n",
    "        # find 1s in the genotype\n",
    "        clade = np.where(genotype == 1)[0]\n",
    "        marker_gene_clades[gene] = set(clade)\n",
    "\n",
    "# marker_gene_clades['INSL4']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>Otherinfo894</th>\n",
       "      <th>Otherinfo895</th>\n",
       "      <th>Otherinfo896</th>\n",
       "      <th>Otherinfo897</th>\n",
       "      <th>Otherinfo898</th>\n",
       "      <th>Otherinfo899</th>\n",
       "      <th>Otherinfo900</th>\n",
       "      <th>Otherinfo901</th>\n",
       "      <th>Otherinfo902</th>\n",
       "      <th>Otherinfo903</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10:100129846:C:A</th>\n",
       "      <td>10</td>\n",
       "      <td>100129846</td>\n",
       "      <td>100129846</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOXL4;PYROXD2</td>\n",
       "      <td>dist=101895;dist=13479</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:20,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100185101:A:G</th>\n",
       "      <td>10</td>\n",
       "      <td>100185101</td>\n",
       "      <td>100185101</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>HPS1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:40,7,1</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>1/1:40,7,1</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100209326:T:C</th>\n",
       "      <td>10</td>\n",
       "      <td>100209326</td>\n",
       "      <td>100209326</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>LOC101927278</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:100283942:C:T</th>\n",
       "      <td>10</td>\n",
       "      <td>100283942</td>\n",
       "      <td>100283942</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>intronic</td>\n",
       "      <td>HPSE2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0,4,13</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:10086005:G:T</th>\n",
       "      <td>10</td>\n",
       "      <td>10086005</td>\n",
       "      <td>10086005</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC101928272;LINC02670</td>\n",
       "      <td>dist=748449;dist=14680</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>0/0:0,2,6</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "      <td>./.:0,0,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 972 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Chr      Start        End Ref Alt    Func.refGene  \\\n",
       "variant_id                                                           \n",
       "10:100129846:C:A  10  100129846  100129846   C   A      intergenic   \n",
       "10:100185101:A:G  10  100185101  100185101   A   G        intronic   \n",
       "10:100209326:T:C  10  100209326  100209326   T   C  ncRNA_intronic   \n",
       "10:100283942:C:T  10  100283942  100283942   C   T        intronic   \n",
       "10:10086005:G:T   10   10086005   10086005   G   T      intergenic   \n",
       "\n",
       "                            Gene.refGene      GeneDetail.refGene  \\\n",
       "variant_id                                                         \n",
       "10:100129846:C:A           LOXL4;PYROXD2  dist=101895;dist=13479   \n",
       "10:100185101:A:G                    HPS1                       .   \n",
       "10:100209326:T:C            LOC101927278                       .   \n",
       "10:100283942:C:T                   HPSE2                       .   \n",
       "10:10086005:G:T   LOC101928272;LINC02670  dist=748449;dist=14680   \n",
       "\n",
       "                 ExonicFunc.refGene AAChange.refGene  ... Otherinfo894  \\\n",
       "variant_id                                            ...                \n",
       "10:100129846:C:A                  .                .  ...    ./.:0,0,0   \n",
       "10:100185101:A:G                  .                .  ...    ./.:0,0,0   \n",
       "10:100209326:T:C                  .                .  ...    ./.:0,0,0   \n",
       "10:100283942:C:T                  .                .  ...   0/0:0,4,13   \n",
       "10:10086005:G:T                   .                .  ...    ./.:0,0,0   \n",
       "\n",
       "                 Otherinfo895 Otherinfo896 Otherinfo897 Otherinfo898  \\\n",
       "variant_id                                                             \n",
       "10:100129846:C:A    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100185101:A:G   1/1:40,7,1    0/0:0,2,6    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100209326:T:C    0/0:0,2,6    ./.:0,0,0    0/0:0,2,6    ./.:0,0,0   \n",
       "10:100283942:C:T    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:10086005:G:T     0/0:0,2,6    0/0:0,2,6    ./.:0,0,0    ./.:0,0,0   \n",
       "\n",
       "                 Otherinfo899 Otherinfo900 Otherinfo901 Otherinfo902  \\\n",
       "variant_id                                                             \n",
       "10:100129846:C:A    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100185101:A:G   1/1:40,7,1    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100209326:T:C    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "10:100283942:C:T    ./.:0,0,0    ./.:0,0,0    0/0:0,2,6    ./.:0,0,0   \n",
       "10:10086005:G:T     ./.:0,0,0    ./.:0,0,0    ./.:0,0,0    ./.:0,0,0   \n",
       "\n",
       "                 Otherinfo903  \n",
       "variant_id                     \n",
       "10:100129846:C:A   1/1:20,3,0  \n",
       "10:100185101:A:G   0/0:0,4,13  \n",
       "10:100209326:T:C    ./.:0,0,0  \n",
       "10:100283942:C:T    0/0:0,2,6  \n",
       "10:10086005:G:T     ./.:0,0,0  \n",
       "\n",
       "[5 rows x 972 columns]"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49\n"
     ]
    }
   ],
   "source": [
    "# find the exonic mutations at the same branch\n",
    "def locate_variant(ix):\n",
    "    variant_id = df_mutations.iloc[ix].snp_id\n",
    "    variant = df_gene_annotation.loc[variant_id]\n",
    "    return variant \n",
    "\n",
    "def find_all_muts_at_same_branch(clade):\n",
    "    all_muts = set()\n",
    "    for site, geno in enumerate(scistree_genotype):\n",
    "        c = set(np.where(geno == 1)[0])\n",
    "        if clade == c:\n",
    "            variant = locate_variant(site)\n",
    "            if variant['Func.refGene'] == 'exonic':\n",
    "                all_muts.add(variant['Gene.refGene'])\n",
    "            # all_muts.add(variant['Gene.refGene'])\n",
    "\n",
    "    return all_muts\n",
    "\n",
    "ancestral_genes = find_all_muts_at_same_branch(marker_gene_clades['TP53'])\n",
    "print(len(ancestral_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PROSER2 244\n",
      "CDC123 891\n",
      "ADAM12 464\n",
      "FAM170B 891\n",
      "EGR2 220\n",
      "TYSND1 891\n",
      "CDH23 234\n",
      "VCL 219\n",
      "PATE2 425\n",
      "NAT10 891\n",
      "INTS5 891\n",
      "PACS1 234\n",
      "BTBD11 213\n",
      "CUX2 145\n",
      "SLC6A12 891\n",
      "PARP11 221\n",
      "SPRYD3 891\n",
      "CCDC77 224\n",
      "PDS5B 891\n",
      "ZC3H13 222\n",
      "FAM124A 891\n",
      "PCDH17 215\n",
      "ZNF839 237\n",
      "KLHL33 424\n",
      "RPGRIP1 891\n",
      "EFS 413\n",
      "SYNE2 188\n",
      "PROX2 218\n",
      "TCL1A 214\n",
      "GANC 891\n",
      "DUOX1 891\n",
      "TLN2 218\n",
      "ZNF710 464\n",
      "CHD2 170\n",
      "SYNM 243\n",
      "SMG1 464\n",
      "ZG16B 203\n",
      "ATXN2L 891\n",
      "RNF40 162\n",
      "GLG1 171\n",
      "ZNF469 891\n",
      "DNAH9 216\n",
      "DHRS13 891\n",
      "PHF12 220\n",
      "ASIC2 235\n",
      "LRRC3C 218\n",
      "TOP2A 207\n",
      "KANSL1 224\n",
      "CACNA1G 184\n",
      "ABCA9 427\n",
      "NEURL4 109\n",
      "TP53 891\n",
      "DNAH2 222\n",
      "EMILIN2 213\n",
      "RPRD1A 416\n",
      "WDR7 199\n",
      "MAP1S 423\n",
      "TLE2 229\n",
      "RYR1 891\n",
      "GRIK5 156\n",
      "ZNF404 200\n",
      "EML2 413\n",
      "GSTM3 145\n",
      "ARNT 413\n",
      "EFNA3 234\n",
      "FCRL3 194\n",
      "RASAL2 891\n",
      "ESRRG 216\n",
      "ACBD3 420\n",
      "HTR1D 243\n",
      "RYR2 891\n",
      "OR2L13 189\n",
      "ARID1A 168\n",
      "GRIK3 198\n",
      "AGRN 195\n",
      "KIZ 243\n",
      "PTPRA 891\n",
      "NCOA6 413\n",
      "PROCR 891\n",
      "LTN1 235\n",
      "GRIK1 210\n",
      "CLDN17 211\n",
      "GAB4 192\n",
      "APOL3 241\n",
      "APOL1 420\n",
      "MYH9 891\n",
      "NPTXR 891\n",
      "SHISAL1 891\n",
      "EDAR 240\n",
      "NEB 131\n",
      "SGO2 891\n",
      "PARD3B 198\n",
      "KANSL1L 891\n",
      "SLC19A3 891\n",
      "ESPNL 226\n",
      "ATAD2B 167\n",
      "KIF1A 891\n",
      "ADGRF3 244\n",
      "BIRC6 162\n",
      "C2orf42 891\n",
      "ADRA2B 150\n",
      "TATDN2 155\n",
      "CASR 241\n",
      "TFDP2 194\n",
      "MED12L 891\n",
      "RARRES1 212\n",
      "XXYLT1 891\n",
      "SEC22C 241\n",
      "GNAT1 891\n",
      "TACR3 241\n",
      "ALPK1 891\n",
      "NOA1 195\n",
      "WDFY3 220\n",
      "ARAP3 239\n",
      "LPCAT1 225\n",
      "MARCOL 891\n",
      "STK10 207\n",
      "MTREX 427\n",
      "LAMA2 413\n",
      "FNDC1 891\n",
      "HDGFL1 215\n",
      "GUCA1ANB 891\n",
      "SRRT 891\n",
      "RINT1 107\n",
      "NRCAM 234\n",
      "PPP1R3A 238\n",
      "FOXP2 891\n",
      "ASB15 201\n",
      "TMEM213 242\n",
      "XRCC2 891\n",
      "SUGCT 891\n",
      "ABCA13 227\n",
      "AKAP9 190\n",
      "ZHX1 190\n",
      "FAM83H 891\n",
      "NPM2 220\n",
      "PBK 425\n",
      "HTRA4 891\n",
      "HOOK3 197\n",
      "PREX2 891\n",
      "ZNF704 891\n",
      "PSKH2 167\n",
      "GEM 162\n",
      "TSPYL5 205\n",
      "CCDC180 234\n",
      "GRIN3A 424\n",
      "PTPN3 241\n",
      "SARDH 891\n",
      "FBXW5 891\n",
      "INSL4 215\n",
      "KANK1 891\n",
      "ECM2 891\n",
      "PCYT1B 891\n",
      "PBDC1 206\n"
     ]
    }
   ],
   "source": [
    "# find cluster-specific mutations\n",
    "def find_cluster_genes(min_leaf=100):\n",
    "    clade_genes = {}\n",
    "    for i, genotype in enumerate(scistree_genotype):\n",
    "        num_leaf = sum(genotype)\n",
    "        if num_leaf > min_leaf:\n",
    "            variant = locate_variant(i)\n",
    "            if variant['Func.refGene'] == 'exonic':\n",
    "                clade_genes[variant['Gene.refGene']] = num_leaf\n",
    "    return clade_genes\n",
    "\n",
    "cluster_genes = find_cluster_genes()\n",
    "for gene in cluster_genes:\n",
    "    print(gene, cluster_genes[gene])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2861.49"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# time (secs) 47min 41.49s\n",
    "time = 47 * 60 + 41.49\n",
    "time\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tree Vis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "pptree = popgen.utils.build_perfect_phylogeny(scistree_genotype.T)\n",
    "df_cell_meta.set_index('id', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use iTOL webserver to visualize the tree\n",
    "# https://itol.embl.de/\n",
    "# generate annotation file\n",
    "# 1-indexed cell ids\n",
    "import seaborn as sns\n",
    "palette = sns.color_palette(\"tab10\").as_hex()\n",
    "clones = df_cell_meta['clone_id'].unique()\n",
    "clone_colors = {clone: palette[ix] for ix, clone in enumerate(clones)}\n",
    "# define my color palette\n",
    "a = sns.color_palette(\"coolwarm\", 18).as_hex()\n",
    "b = sns.light_palette(\"seagreen\").as_hex()\n",
    "clone_colors = {10: a[0],\n",
    "                11: a[1],\n",
    "                1: a[5],\n",
    "                8: a[6],\n",
    "                0: b[1],\n",
    "                17:b[4],\n",
    "                6:a[13],\n",
    "                16:a[15],\n",
    "                14:a[17]\n",
    "}\n",
    "\n",
    "sample_colors = {'SA1090': palette[0], \n",
    "                 'SA921': palette[1],\n",
    "                 'SA922': palette[2],}\n",
    "def generate_itol_annotation_clone(pptree, df_cell_meta):\n",
    "    with open(f'{prefix}.clone.itol.txt', 'w') as f:\n",
    "        f.write('DATASET_COLORSTRIP\\n')\n",
    "        f.write('SEPARATOR COMMA\\n')\n",
    "        f.write('DATASET_LABEL,OV2295\\n')\n",
    "        f.write('COLOR,#ff0000\\n')\n",
    "        f.write('DATA\\n')\n",
    "        for leaf in pptree.get_leaves():\n",
    "            clone = df_cell_meta.loc[leaf, 'clone_id']\n",
    "            f.write(f'{leaf+1},{clone_colors[clone]},clone{clone}\\n')\n",
    "\n",
    "def generate_itol_annotation_sample(pptree, df_cell_meta):\n",
    "    with open(f'{prefix}.sample.itol.txt', 'w') as f:\n",
    "        f.write('DATASET_COLORSTRIP\\n')\n",
    "        f.write('SEPARATOR COMMA\\n')\n",
    "        f.write('DATASET_LABEL,OV2295\\n')\n",
    "        f.write('COLOR,#ff0000\\n')\n",
    "        f.write('DATA\\n')\n",
    "        for leaf in pptree.get_leaves():\n",
    "            sample = df_cell_meta.loc[leaf, 'sample_id']\n",
    "            f.write(f'{leaf+1},{sample_colors[sample]},{sample}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_itol_annotation_clone(pptree, df_cell_meta)\n",
    "generate_itol_annotation_sample(pptree, df_cell_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({10: '#4a63d3',\n",
       "  11: '#5a78e4',\n",
       "  1: '#a3c2fe',\n",
       "  8: '#b6cefa',\n",
       "  0: '#c5decf',\n",
       "  17: '#54a075',\n",
       "  6: '#f59f80',\n",
       "  16: '#e57058',\n",
       "  14: '#c73635'},\n",
       " {'SA1090': '#1f77b4', 'SA921': '#ff7f0e', 'SA922': '#2ca02c'})"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clone colors\n",
    "clone_colors, sample_colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "clone_id\n",
       "0     145\n",
       "1      99\n",
       "6      96\n",
       "8     110\n",
       "10    126\n",
       "11     92\n",
       "14     86\n",
       "16     62\n",
       "17     75\n",
       "dtype: int64"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find number of cells in each clone \n",
    "df_clones.groupby('clone_id').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'((((3 10,4 11)2,(6 1,7 8)5)1,((10 0,11 17)9,((14 6,15 16)13,16 14)12)8)0);'"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('../ov2295_tree.pickle', 'rb') as f:\n",
    "    clonal_tree = pickle.load(f)\n",
    "clonal_tree.newick_str()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SA1090', 'SA921', 'SA922'], dtype=object)"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cell_meta['sample_id'].unique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "work",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
